pinot-controller/src/main/java/org/apache/pinot/controller/workload/QueryWorkloadManager.java

```diff
@@ -170,34 +312,182 @@ public Map<String, InstanceCost> getWorkloadToInstanceCostFor(String instanceNam
+        _propagationSchemeProvider.getPropagationScheme(nodeConfig.getPropagationScheme().getPropagationType());
     return propagationScheme.resolveInstances(nodeConfig);
   }
+
+  /**
+   * For propagation entities with empty cpu or memory cost, distribute the remaining cost evenly among them.
+   * If all entities have defined costs, we do nothing.
+   *
+   * @param nodeConfig The node config containing the propagation entities to check and distribute costs for.
+   */
+  private void checkAndDistributeEmptyPropagationEntitiesEvenly(NodeConfig nodeConfig) {
+    List<PropagationEntity> propagationEntities = nodeConfig.getPropagationScheme().getPropagationEntities();
+    int emptyEntitiesCount = 0;
+    int definedEntitiesCount = 0;
+    for (PropagationEntity entity : propagationEntities) {
+      if (entity.getCpuCostNs() == null || entity.getMemoryCostBytes() == null) {
+        emptyEntitiesCount++;
+      } else {
+        definedEntitiesCount++;
+      }
+    }
+    if (definedEntitiesCount > 0 && emptyEntitiesCount > 0) {
+      String errorMsg = String.format("Mixed defined and empty costs in propagation entities is not supported. "
+          + "NodeConfig: %s", nodeConfig);
+      LOGGER.error(errorMsg);
+      throw new RuntimeException(errorMsg);
+    }
+    if (emptyEntitiesCount == propagationEntities.size()) {
+      // All entities have empty costs - distribute total budget evenly
+      long totalCpuCostNs = nodeConfig.getEnforcementProfile().getCpuCostNs();
+      long totalMemoryCostBytes = nodeConfig.getEnforcementProfile().getMemoryCostBytes();
+
+      int numEntities = propagationEntities.size();
+      long shareCpuCostNs = totalCpuCostNs / numEntities;
+      long shareMemoryCostBytes = totalMemoryCostBytes / numEntities;
+
+      for (PropagationEntity entity : propagationEntities) {
+        entity.setCpuCostNs(shareCpuCostNs);
+        entity.setMemoryCostBytes(shareMemoryCostBytes);
+      }
+      LOGGER.info("Distributed costs evenly across {} entities: CPU={}ns, Memory={}bytes per entity",
+          numEntities, shareCpuCostNs, shareMemoryCostBytes);
+    }
+  }
+
+  /**
+   * Sends the given map of {@link QueryWorkloadRefreshMessage} to their respective instances asynchronously.
+   *
+   * <p>
+   * Each message is sent in its own asynchronous task, and the method waits for all tasks to complete
+   * with a timeout of 60 seconds per instance. Success and failure counts are logged.
+   * </p>
+   *
+   * @param instanceToRefreshMessageMap A map from instance name to the {@link QueryWorkloadRefreshMessage} to send.
+   */
+  public void sendQueryWorkloadRefreshMessage(Map<String, QueryWorkloadRefreshMessage> instanceToRefreshMessageMap) {
+    ClusterMessagingService messagingService = _pinotHelixResourceManager.getHelixZkManager().getMessagingService();
+    List<CompletableFuture<Boolean>> futures = instanceToRefreshMessageMap.entrySet().stream()
```

Comment:
You had a valid reason for handling the message sending logic this - latency for queuing each task. Can you please add a detailed comment here explaining why you decided to do it this way? 

An alternative here is to rely on Server API calls to send this detail to the server. Please create a ticket for yourself and let's explore this as a part of the perf benchmarking.